spec/thawed_mock.rb

```diff
@@ -0,0 +1,112 @@
+# frozen_string_literal: true
+
+module ThawedMock
+  OBJECTS = {}
+
+  def self.allow_mocking(obj, *methods)
+    mod, mock = OBJECTS[obj]
+    unless mod
+      mod = Module.new
+      mock = Object.new
+
+      obj.singleton_class.prepend(mod)
+      OBJECTS[obj] = [mod, mock].freeze
+    end
+
+    methods.each do |method|
+      original_method = obj.method(method)
+
+      mock.define_singleton_method(method) do |*a, **kw, &b|
+        original_method.call(*a, **kw, &b)
+      end
+
+      mod.define_method(method) do |*a, **kw, &b|
+        mock.send(method, *a, **kw, &b)
+      end
+    end
+  end
+
+  module ExpectOverride
+    [:allow, :expect].each do |method|
+      define_method(method) do |obj = nil, &block|
+        return super(&block) if obj.nil? && block
+
+        _, mock = OBJECTS[obj]
+        super(mock || obj, &block)
+      end
+    end
+  end
+  RSpec::Core::ExampleGroup.prepend(ExpectOverride)
```

Comment:
Nice, I was thinking of messing with expect to do this, but honestly my Ruby language lawyering is not good enough, I use it like a C programmer who was taught lambda calculus in scheme that one time.